src/DI/ContainerFactory.php

```diff
@@ -110,44 +110,54 @@ private function loadClass()
 	{
 		$key = md5(serialize(array($this->config, $this->configFiles, $this->class, $this->parentClass)));
 		$file = "$this->tempDirectory/$key.php";
-
-		if (!$this->autoRebuild && (@include $file) !== FALSE) { // @ - file may not exist
+		if (!$this->isExpired($file) && (@include $file) !== FALSE) {
 			return;
 		}
 
+		if (!is_dir($this->tempDirectory)) {
+			@mkdir($this->tempDirectory); // @ - directory may already exist
+		}
+
 		$handle = fopen("$file.lock", 'c+');
-		if (!$handle) {
-			throw new Nette\IOException("Unable to open or create file '$file.lock'.");
+		if (!$handle || !flock($handle, LOCK_EX)) {
+			throw new Nette\IOException("Unable to acquire exclusive lock '$file.lock'.");
 		}
 
-		if ($this->autoRebuild) {
-			flock($handle, LOCK_SH);
-			foreach ((array) @unserialize(file_get_contents("$file.meta")) as $f => $time) { // @ - file may not exist
-				if (@filemtime($f) !== $time) { // @ - stat may fail
-					@unlink($file); // @ - file may not exist
-					break;
-				}
+		if (!is_file($file) || $this->isExpired($file)) {
+			$this->dependencies = array();
+			$code = $this->generateCode();
+			$meta = serialize(array_combine($this->dependencies, @array_map('filemtime', $this->dependencies)));
+			if (file_put_contents("$file.tmp", $meta) !== strlen($meta) || !rename("$file.tmp", "$file.meta")) {
+				@unlink("$file.tmp"); // @ - file may not exist
+				throw new Nette\IOException("Unable to create file '$file.meta'.");
+			}
+			if (file_put_contents("$file.tmp", $code) !== strlen($code) || !rename("$file.tmp", $file)) {
+				@unlink("$file.tmp"); // @ - file may not exist
+				throw new Nette\IOException("Unable to create file '$file'.");
 			}
 		}
 
-		if (!is_file($file)) {
-			flock($handle, LOCK_EX);
-			if (!is_file($file)) {
-				$this->dependencies = array();
-				$code = $this->generateCode();
-				if (file_put_contents("$file.tmp", $code) !== strlen($code) || !rename("$file.tmp", $file)) {
-					@unlink("$file.tmp"); // @ - file may not be created
-					throw new Nette\IOException("Unable to create file '$file'.");
-				}
-				$tmp = array();
-				foreach ($this->dependencies as $f) {
-					$tmp[$f] = @filemtime($f); // @ - stat may fail
+		if ((@include $file) === FALSE) { // @ - error escalated to exception
+			throw new Nette\IOException("Unable to include '$file'.");
+		}
+		flock($handle, LOCK_UN);
+	}
```

Comment:
IMHO they can't be delete, I think I even tried it and it caused problems. I don't have to retest it or think about it more thoroughly now.